services/src/main/scala/cromwell/services/healthmonitor/HealthMonitorServiceActor.scala

```diff
@@ -0,0 +1,131 @@
+import scala.language.postfixOps
+import HealthMonitorServiceActor._
+import com.typesafe.config.Config
+import cromwell.services.ServiceRegistryActor.ServiceRegistryMessage
+import net.ceedubs.ficus.Ficus._
+
+/**
+  * A trait which provides a health monitoring service. Implementers are expected to provide a collection of functions
+  * which will each check some underlying service on which Cromwell depends, for instance connectivity to the database
+  * or DockerHub. At a regular interval the health monitor actor will run each of these functions and keep their
+  * current state in a cache.
+  *
+  * This actor may also be interrogated for the current status of the monitored subsystems. The response will indicate
+  * if all of the subsystems are deemed to be healthy, and if this is not true subsystem-specific messages will be
+  * returned to the client.
+  */
+trait HealthMonitorServiceActor extends Actor with LazyLogging with Timers {
+  val serviceConfig: Config
+  def subsystems: Set[MonitoredSubsystem]
+
+  implicit val ec: ExecutionContext = context.system.dispatcher
+
+  lazy val futureTimeout: FiniteDuration = serviceConfig.as[Option[FiniteDuration]]("services.HealthMonitor.check-timeout").getOrElse(DefaultFutureTimeout)
+  lazy val staleThreshold: FiniteDuration = serviceConfig.as[Option[FiniteDuration]]("services.HealthMonitor.status-ttl").getOrElse(DefaultStaleThreshold)
+
+  override def preStart(): Unit = {
+    val sweepTime = serviceConfig.as[Option[FiniteDuration]]("services.HealthMonitor.check-refresh-time").getOrElse(DefaultSweepTime)
+    timers.startPeriodicTimer(CheckTickKey, CheckAll, sweepTime)
+    logger.info("Starting health monitor with the following checks: " + subsystems.map(_.name).mkString(", "))
+  }
+
+  /**
+    * Contains each subsystem status along with a timestamp of when the entry was made so we know when the status
+    * goes stale. Initialized with unknown status.
+    */
+  private var statusCache: Map[MonitoredSubsystem, CachedSubsystemStatus] = {
+    val now = System.currentTimeMillis
+    subsystems.map((_, CachedSubsystemStatus(UnknownStatus, now))).toMap
+  }
+
+  override def receive: Receive = {
+    case CheckAll =>
+      logger.debug("Checking status on all subsystems")
+      subsystems.foreach(checkSubsystem)
+    case Store(subsystem, status) => store(subsystem, status)
+    case GetCurrentStatus => sender ! getCurrentStatus
+    case ShutdownCommand => context.stop(self) // Not necessary but service registry requires it. See #2575
+  }
+
+  private def checkSubsystem(subsystem: MonitoredSubsystem): Unit = {
+    val result = subsystem.check()
+    val errMsg = s"Timed out after ${futureTimeout.toString} waiting for a response from ${subsystem.toString}"
+    result.withTimeout(futureTimeout, errMsg, context.system.scheduler)
+      .recover { case NonFatal(ex) =>
+        failedStatus(ex.getMessage)
+      } map {
+      Store(subsystem, _)
+    } pipeTo self
```

Comment:
ToL curious why not call `store` here as opposed to this `pipeTo`?

Reply:
what happens if the failure is `Fatal`?

Reply:
The real answer is this block was taken verbatim from Rawls and any answer I'd give would be guessing on the intent.

Reply:
@mcovarr That's a good question. `pipeTo` will send a `Status.Failure` message which is not handled as far as I can tell. I don't *think* there's any underlying akka magic for that. Let me double check w/ @rtitle that there's not some slick way this is really being handled that I missed.

Reply:
There doesn't seem to be anything in `receive` that would deal with an Akka `Failure`, and unless I've missed something that seems like a situation that could happen.

Reply:
@mcovarr That was exactly what I was just saying, but using different words. The other bit I was getting at is that it's possible there was another part to this which I missed and didn't bring across to Cromwell.

Reply:
Have chatted with @rtitle - will be setting up an experiment to see what happens. I'd like to think that this is a non-issue as this feels like a real easy trap to fall into with pipeTo but early investigation was unclear

Reply:
Confirmed that nothing good happens here

Reply:
@danbills - @rtitle pointed out that the reason for the `pipeTo` was so that actor state wasn't being modified from inside a `Future` which IMO is a very good reason

Reply:
That's an excellent reason which ought to be immortalized in a comment. 🙂

Reply:
@mcovarr A year ago I'd have said that that should be so obvious that no one would ever question it. And then this thread happened.

Reply:
@mcovarr So it turns out that my test was wrong. I was constructing a hard error with `Future.failed` which doesn't get caught by `NonFatal()`. With a true fatal error this does the rights thing and shuts things down.

That said ..... if a check was doing this intentionally (e.g. `Future.failed(new OutOfMemoryError())` it wouldn't get caught. I'd argue that this isn't worth the extra hijinks to protect against since even if this were to happen it'd be programmer hijinks and not a true fatal error.